Auth features for server

bryevdv opened the pull request:
- [x] issues: fixes #9140
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

This is WIP, would like to get some early feed back on how the usage feels.

You can now supply an auth modue, e.g. 
```

import tornado
from tornado.web import RequestHandler

# could also define async_get_user
def get_user(request_handler):
    return request_handler.get_secure_cookie("user")

# could also define get_login_url func (but give up LoginHandler)
login_url = "/login"

# optional login page at login_url
class LoginHandler(RequestHandler):

    def get(self):
        try:
            errormessage = self.get_argument("error")
        except:
            errormessage = ""
        self.render("login.html", errormessage=errormessage)

    def check_permission(self, password, username):
        if username == "bokeh" and password == "bokeh":
            return True
        return False

    def post(self):
        username = self.get_argument("username", "")
        password = self.get_argument("password", "")
        auth = self.check_permission(password, username)
        if auth:
            self.set_current_user(username)
            self.redirect(self.get_argument("next", u"/"))
        else:
            error_msg = u"?error=" + tornado.escape.url_escape("Login incorrect")
            self.redirect(login_url + error_msg)

    def set_current_user(self, user):
        if user:
            self.set_secure_cookie("user", tornado.escape.json_encode(user))
        else:
            self.clear_cookie("user")
```
and a `login.html` template, and you can only get to the app if you pass login. 


Some notes: 

* always have to supply `--auth-module` separately, i.e. can't make `auth.py` in a directoy app work because there could be multiple directory apps and that makes no sense (also other reasons)
* `LoginHandler` is optional, "serious" deploys would redirect to some external login probably
* Since any cookie get/set happens in the auth module, users can control max age and expiry there themselves when they call those funcs
* Do we need to support a `LogoutHandler`? Presumably apps could clear the cookie themselved from a Button push, etc. Maybe later?

Initial thoughts, @bokeh/dev ?

bryevdv commented:
Here is a simple example:

![ScreenFlow](https://user-images.githubusercontent.com/1078448/63644958-85ea2c80-c6a9-11e9-8274-38bd18be46d5.gif)

birdsarah commented:
Sorry I'm being slow, can you spell out the rest of the incantation that would mean this is all hooked up?

bryevdv commented:
Sorry, run like this:
```
BOKEH_COOKIE_SECRET='muy super secret' bokeh serve --auth-module=/Users/bryan/tmp/auth.py --show sliders.py
```

I used this template:
```
<div id="main-container">
    <div id="main">
        <div id="login-form">
            <form action="/login" method="post" id="login_form">
                <fieldset>
                    <label for="username">Username</label>
                    <input autocapitalize="off" autocorrect="off" class="text-input" id="username" name="username" tabindex="1" type="text" value="">
                </fieldset>

                <fieldset>
                    <label for="password">Password</label>
                    <input class="text-input" id="password" name="password" tabindex="2" type="password" value="">
                </fieldset>
                <fieldset>
                    <span class="errormessage">{{errormessage}}</span>
                </fieldset>

                <div id="form_btn">
                    <input id="signin-btn" class="btn btn-blue" type="submit" value="Sign In" tabindex="3">
                </div>
            </form>
        </div>
    </div>
</div>
```

p-himik commented on bokeh/server/tornado.py:
```diff
@@ -249,6 +280,34 @@ def __init__(self,
+        if login_handler and not issubclass(login_handler, RequestHandler):
+            raise ValueError("LoginHandler must be a Tornado RequestHandler")
+        # This just catches some common cases up front, let tornado barf on any others
+        if login_handler and login_url.startswith("http") and login_url.startswith("//"):
```
Always `False`.

bryevdv replied:
Thank you!

p-himik replied:
Also technically a legitimate relative URL could start with `http` but I guess it's not a big deal.

bryevdv replied:
No it's a fair observation, I will tighten it up. I'm not worried about false negatives here since Tornado will catch the issue if we don't, but should try to avoid any false positives.

birdsarah commented:
This looms great. Just some clear docs like the other stuff you've done for server. The only bit that I'm left wondering is which class is bokeh looking for - anything that overrides the requesthandler? Or is it the loginhandler class specifically which needs to override the requesthandler.

The enterprise I imagine is the bokeh server being behind the corporate ldap/oauth where you go to the bokeh page, get redirected to oauth, and then get let in if valid, in that case no template would be required its just routing and checking.

The other scenario I imagine is where your bokeh app is embedded in site that has login. In that case the bokeh plot would be an embedded frame that would gate until author handling / checking was complete.

In both of these scenarios the login template itself wouldn't be used. But I think its great to include your scenario too wherethe only server is bokeh / tornado and it serves up the login.

birdsarah commented:
Maybe this leads to a more general 'hookig into tornado', I'm not sure.

bryevdv commented:
@p-himik I believe I have addressed your change in 851f00c

@birdsarah @hhuuggoo I think something needs to be done re: logging out. I am just not sure what exactly the best workflow is. I think it would be simple to add a `logout_url` or `LogoutHandler` analagous to the login one. That handler could clear cookies or whatever, etc. Then a  `js_on_click` or session timeout callback could redirect to that URL. Does that sound reasonable?

birdsarah commented:
> I think something needs to be done re: logging out. I am just not sure what exactly the best workflow is. I think it would be simple to add a logout_url and optional LogoutHandler analagous to the login case. That handler could clear cookies or whatever, etc. 

Good point. Yes. I don't remember ever worrying about logging people out specifically. Typically you just set the cookie expiry to be what you want. I don't know what would happen to bokeh once the cookie expired. Regardless the analagous logout hook would be great, and presumably it's easy.

bryevdv commented:
Ok I added the logout url/handler support:

![ScreenFlow](https://user-images.githubusercontent.com/1078448/64478839-3bfe4d80-d163-11e9-877f-51f793057a42.gif)

Imagine if we had added this three years ago 🤦‍♂ 

I've been in a bit of a slump the last two weeks but I am going to do my best to get these things done this weekend:

* Factor all the auth things from the auth module into an `AuthProvider` class. There are lots of bits and they are all currently smeared all over the tornado handler
* Add tests (first task will help with this)
* User guide docs

Hoping to get some final thoughts/review and merge early next week.

bryevdv commented:
OK one more small feature, `--enable-xsrf-cookies` flag to turn to Tornado's XSRF protection.

bryevdv commented:
@bokeh/dev I am only going to add test from here out, marking this `ready` for review

p-himik commented on bokeh/command/subcommands/serve.py:
```diff
@@ -388,10 +459,10 @@
+        metavar = 'CONFIG',
+        type    = str,
+        help    = "Use a YAML config file for settings",
+        default = None,
```
Judging by this PR, the style for keyword arguments' values alignment is different throughout Bokeh codebase. And I've never seen such alignment in any Python project before, although that might be because of my limited exposure to other Python projects. Is there any prettifier config or documentation that makes it clear that this style is preferred?

bryevdv replied:
No presently, ideally we can just switch to black maybe at some point and forget about debates. I would never write kwargs in an actual function call this way, but although these are ultimately used that way, here they are more like data literals and that is the case where this is my personal pref (see also: lining up dict literal values)

bryevdv replied:
In this case I mostly just wanted to make things consistent even within the same file

p-himik commented on bokeh/server/auth_provider.py:
```diff
@@ -0,0 +1,318 @@
+
+    def __init__(self, module_path):
+        if not isfile(module_path):
+            raise ValueError()
```
Maybe a message would make sense.

bryevdv replied:
I meant to go back and addd one but forgot, thank you

bryevdv replied:
https://github.com/bokeh/bokeh/pull/9190/commits/d06f2e2def0a503e11a51f6e372d8856ca5410b9

p-himik commented on bokeh/server/auth_provider.py:
```diff
@@ -0,0 +1,318 @@
+        bool
+
+    '''
+    return not (url.startswith("http:") or url.startswith("https:") or url.startswith("//"))
```
Just realized - could be simplified to `return not url.startswith(("http://", "https://", "//"))`. Even on Python 2.7.

bryevdv replied:
https://github.com/bokeh/bokeh/pull/9190/commits/5a76573f27f177d24981570445bee59451c8f43b

bryevdv commented:
#bokeh/dev Barring serious issues raised, I intend to merge this tomorrow.

philippjfr commented on sphinx/source/docs/user_guide/server.rst:
```diff
@@ -1106,6 +1111,91 @@ here:
+authorization. However, the Bokeh server can be configured with and "Auth
+Provider" that hooks in to Tornado's underlying capabilities. For background
+information, see the Tornado docs for `Authentication and security`_. The rest
+of this section assume some familiarity with that material.
```
```suggestion
of this section assumes some familiarity with that material.
```

philippjfr commented on sphinx/source/docs/user_guide/server.rst:
```diff
@@ -1106,6 +1111,91 @@ here:
+    def get_user(request_handler):
+        pass
+
+    async def get_user_aync(request_handler):
```
```suggestion
    async def get_user_async(request_handler):
```

philippjfr commented on sphinx/source/docs/user_guide/server.rst:
```diff
@@ -1106,6 +1111,91 @@ here:
+``LoginHandler`` when ``get_url_function`` is defined.
+
+Analogous to the login options, optional ``logout_url`` and ``LogoutHandler``
+values may be define and endopoint for logging users out.
```
```suggestion
values may be define an endpoint for logging users out.
```

philippjfr commented on bokeh/settings.py:
```diff
@@ -568,6 +583,17 @@ def dev(self):
     """)
 
+    xsrf_cookies = PrioritizedSetting("xsrf_cookies", "BOKEH_XSRF_COOKIES", default=False, convert=convert_bool, help="""
+    Whether to enable Tornado XSRF cookie protetion on the Bokeh server. This
```
```suggestion
    Whether to enable Tornado XSRF cookie protection on the Bokeh server. This
```

philippjfr commented on bokeh/command/subcommands/serve.py:
```diff
@@ -246,6 +246,76 @@
+    def get_user(request_handler):
+        pass
+
+    async def get_user_aync(request_handler):
```
```suggestion
    async def get_user_async(request_handler):
```

philippjfr commented on bokeh/command/subcommands/serve.py:
```diff
@@ -246,6 +246,76 @@
+``LoginHandler`` when ``get_url_function`` is defined.
+
+Analogous to the login options, optional ``logout_url`` and ``LogoutHandler``
+values may be define and endopoint for logging users out.
```
```suggestion
values may be define an endpoint for logging users out.
```

philippjfr commented on bokeh/server/auth_provider.py:
```diff
@@ -0,0 +1,318 @@
+        ''' A function that computes a URL to redirect unathenticated users
+        to for login.
+
+        This proprty may return None, if a ``login_url`` is supplied
```
```suggestion
        This property may return None, if a ``login_url`` is supplied
```

philippjfr commented on bokeh/server/auth_provider.py:
```diff
@@ -0,0 +1,318 @@
+
+    @property
+    def get_user_async(self):
+        ''' An aysnc function to get the current authenticated user.
```
```suggestion
        ''' An async function to get the current authenticated user.
```

philippjfr commented on bokeh/server/auth_provider.py:
```diff
@@ -0,0 +1,318 @@
+
+    @property
+    def login_url(self):
+        ''' A URL to redirect unathenticated users to for login.
```
```suggestion
        ''' A URL to redirect unauthenticated users to for login.
```

jbednar commented on bokeh/command/subcommands/serve.py:
```diff
@@ -246,6 +246,76 @@
+~~~~~~~~~~~~~~~~~~~~~~
+
+The Bokeh server can be configured to only allow connections in case there is
+a properly authenticed user. This is accomplished by providing the path to
```
```suggestion
a properly authenticated user. This is accomplished by providing the path to
```

jbednar commented on bokeh/command/subcommands/serve.py:
```diff
@@ -246,6 +246,76 @@
+authentication will be required to access Bokeh server endpoints.
+
+.. warning::
+    The contents of the auth module will be executed!
```
Can you explain the implications here, i.e. why that's a warning? Maybe:

```suggestion
    The contents of the auth module will be executed, so you should ensure that this module is trustworthy and only includes code you are comfortable with.
```

jbednar commented on bokeh/command/subcommands/serve.py:
```diff
@@ -450,119 +521,141 @@ class Serve(Subcommand):
+            metavar = 'AUTH_MODULE',
+            action  = 'store',
+            default = None,
+            help    = 'Absolute path to a Python modules that implements auth hooks',
```
```suggestion
            help    = 'Absolute path to a Python module that implements auth hooks',
```

jbednar commented on bokeh/command/subcommands/tests/test_serve.py:
```diff
@@ -144,119 +146,141 @@ def test_args():
+            metavar = 'AUTH_MODULE',
+            action  = 'store',
+            default = None,
+            help    = 'Absolute path to a Python modules that implements auth hooks',
```
```suggestion
            help    = 'Absolute path to a Python module that implements auth hooks',
```

jbednar commented on bokeh/settings.py:
```diff
@@ -413,6 +413,15 @@ def dev(self):
+    the Bokeh server.
+
+    .. warning::
+        The contents of this module will be executed!
```
(Same as above comment)

jbednar commented on sphinx/source/docs/user_guide/server.rst:
```diff
@@ -8,6 +8,10 @@ Running a Bokeh Server
 Purpose
 -------
 
+The purpose of the Bokeh server is to make it easy for python users to create
```
```suggestion
The purpose of the Bokeh server is to make it easy for Python users to create
```

jbednar commented on sphinx/source/docs/user_guide/server.rst:
```diff
@@ -1106,6 +1111,91 @@ here:
+~~~~~~~~~~~~~~
+
+The Bokeh server itself does not have any facilities for authentication or
+authorization. However, the Bokeh server can be configured with and "Auth
```
```suggestion
authorization. However, the Bokeh server can be configured with an "Auth
```

jbednar commented on sphinx/source/docs/user_guide/server.rst:
```diff
@@ -1106,6 +1111,91 @@ here:
+'''''''''''
+
+The Bokeh server can be configured to only allow connections in case there is
+a properly authenticed user. This is accomplished by providing the path to
```
```suggestion
a properly authenticated user. This is accomplished by providing the path to
```

jbednar commented on sphinx/source/docs/user_guide/server.rst:
```diff
@@ -1232,6 +1322,22 @@ Django or whatever tool is in use).
+
+    {% module xsrf_form_html() %}
+
+to all HTML form submissions templates. For full details, see the Tornado
```
```suggestion
to all HTML form submission templates. For full details, see the Tornado
```

bryevdv commented:
Ok I am going to go ahead and merge this. Release is still at least a few weeks out so there is certainly time for any follow on work.
